### PR TITLE
[FIX] stock: be even more precise to not loose data in case needed

### PR DESCRIPTION
**Pull request by wtaferner**
**Description of the issue/feature this PR addresses:**
Fine-tuning and suggesting. As this was hitting hard after https://github.com/odoo/odoo/commit/4a5f07654f6d4252f73253d26979cf97bbf7f11b and before https://github.com/odoo/odoo/commit/a29ec8008201ed7ffaed32fd3960cd5ea76c5ff6 I had some time to review and think about the solution finally provided and I suggest to even improve the current wiping out to be more precise by splitting the condition and do not take a boolean as value for comparison

**Current behavior before PR:**
It might theoretically happen that information is lost which is of interest for some use cases.

**Desired behavior after PR is merged:**
Only really ambivalent information is cleaned up on the picking.

@sle-odoo 
Thank you for fixing quick the oversight. I would like to suggest this one in order to further improve. May I ask you to review and merge or explain in case you think the improvement is not correct.

Not sure about the origin...would it actually make sense to join them as you stated? I might not have the use cases in mind you have.

@jco-odoo @amoyaux @pimodoo 
If you are interested, feel free to comment and review :-)

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by wtaferner** [comment]
I just extended this topic by two other related commits and use cases from the wild (warehouses and fulfilling users out there).

The one would have made the removal of the origin and the partner not soo bad as it would still hold the correct reference in case we only have one sale order or purchase order related to a picking.

The other one was the need to get real quick whether a picking can be fully fulfilled before starting. Of course the fulfilling person will always get to know whether it was a complete or a partial fulfillment when he/she is asked to create a backorder or not, but it is good to have the information upfront.

**Comment by wtaferner** [comment]
@lap-odoo 
As this is also functional now, feel free to get involved. Also choose to cherry-pick for master if you do not want to merge all three commits in stable.

If any of you have request for changes or improvement before you want to merge, I will be delighted to provide them as this removes the size of my customization ;-)

**Comment by sle-odoo** [comment]
@wtaferner Hi, we'll have a look on Monday. Thanks for your contribution. A quick remark though, about the last point:
>the need to get real quick whether a picking can be fully fulfilled before starting

can't you simply use the picking state? If a picking has a deliver partially strategy, he'll be "ready" directly when any of its move is [partially ]reserved. If he has a deliver all at once strategy, he'll only be "ready" when all its stock moves are fully reserved.

**Comment by wtaferner** [comment]
@sle-odoo 
Great, I really appreciate that...I have put quite some thoughts.

About the picking state...well, the policy would not fit all use cases and sometimes people need to decide (based on what kind of goods are shipped together or not), but should be aware upfront if it makes sense to invest the time and fulfill first what is easy and without doubt.

By the way, I think there is a bug in the picking state computation...if there is no reservation it should never be ready, but waiting...I have quite some "empty" ToDo transfers...I might investigate in the next weeks.

File: addons/stock/models/stock_picking.py [hunk 1]
```diff
@@ -295,11 +295,23 @@ class Picking(models.Model):
     product_id = fields.Many2one('product.product', 'Product', related='move_lines.product_id')
     show_operations = fields.Boolean(compute='_compute_show_operations')
     show_lots_text = fields.Boolean(compute='_compute_show_lots_text')
+    partial_delivery = fields.Boolean(
+        string='Partial Delivery',
+        compute='_compute_partial_delivery',
+    )
```
Comment (wtaferner) [inline addons/stock/models/stock_picking.py]: If this is merged in master we should make it searchable and stored.

**Comment by wtaferner** [comment]
Ok, found the little glitch for the post unreserve, reserve wrong state for the partially available state without reservations which is a state "waiting" instaed of "ready"...just added this commit as well.

**Comment by wtaferner** [comment]
@sle-odoo 
Which Monday did you refer to? ;-) Not all of it is buggy behavior, but some of the commits are needed to be fixed asap.

**Comment by sle-odoo** [comment]
@wtaferner Sorry but we have a lot of work, we try to make the best of our time. I'll try to find some for this pr this week. Don't take this the wrong way but if you want a bug fixed in a timely fashion, you know you can still open an opw.

**Comment by wtaferner** [comment]
@sle-odoo 
Sure, nothing easier than that ;-) 
OPW-783544

I know that you have of tons of work like I do but of course I think these commits have quite some positive impact if I have not overseen something.

**Comment by sle-odoo** [comment]
@wtaferner 
[commit 1](https://github.com/odoo/odoo/pull/20867/commits/9f5139d66dac62e4fbd67db5a91521782651f8e8): can you explain a real life usecase where this matters? i guess it's to keep the partner_id where the procurement grouped pick is only linked to one, but as soon as another partner is linked it'll be wiped out anyway. and when the user sets a procurement group to group some operations, most of the times it's linked to multiple partners + the operator should not care about the partner anyway (batch picking)
[commit 2](https://github.com/odoo/odoo/pull/20867/commits/4fcfcf40dca94160b8640f619420e8f6408e9331): i don't see what we improve here, it'll be filled the same as origin but it's clickable.. not sure we need it. and it should be wiped in _assign_picking the same way origin is
[commit 3](https://github.com/odoo/odoo/pull/20867/commits/0e9e2451f2094280d453ed0442f76220706210f5): can you explain a real use case i can reproduce?  also, you could have unreserved move lines (product_qty is zero)
[commit 4](https://github.com/odoo/odoo/pull/20867/commits/7fac61a1ce7249d827540535fae7dcfc3da3c546): still not convinced the picking state is not enough :) i'll talk about it with @lap-odoo

**Comment by wtaferner** [comment]
@sle-odoo 
[commit 1](https://github.com/odoo/odoo/pull/20867/commits/9f5139d66dac62e4fbd67db5a91521782651f8e8): well, this is simply following the principle...better be explicit than implicit and as your first commit destroyed things your second one might fix it for the case we had but we should not assume that there are no other corner cases out there, so I did improve it for more possible states that nothing is deleted by coincidence...I think there was also a reason why I did it but if it is not written in the PR, it is probably lost...too many things happened in the meantime.

[commit 2](https://github.com/odoo/odoo/pull/20867/commits/4fcfcf40dca94160b8640f619420e8f6408e9331): Well, it is just to give a direct link to the sale order which is missing (the origin is nice, but honestly a click is fast and easy if there is only one relation. So this commit is giving a solution...matter of taste, but I dislike the origin as only source of relation and the chatter is packed with not useless, but flooded messages about reserve, unreserve, reserve, unreserve, etc. of stock move states.

[commit 3](https://github.com/odoo/odoo/pull/20867/commits/0e9e2451f2094280d453ed0442f76220706210f5): This one is simple...you will have a picking in the todo list although there is not a single reservation...this is why I had to fix this for my people in the warehouse...useless information in the ToDo list.

[commit 4](https://github.com/odoo/odoo/pull/20867/commits/7fac61a1ce7249d827540535fae7dcfc3da3c546): not convinced? hmm, you should come to the two warehouses having a lot of information and you will be immediately convinced that this is a life saver ;-)

**Comment by sle-odoo** [comment]
> commit 1: well, this is simply following the principle...better be explicit than implicit and as your first commit destroyed things your second one might fix it for the case we had but we should not assume that there are no other corner cases out there, so I did improve it for more possible states that nothing is deleted by coincidence...I think there was also a reason why I did it but if it is not written in the PR, it is probably lost...too many things happened in the meantime.

I prefer to not make a patch for a potential corner case. And, as i explained, as soon as a different delivery is merged with a different partner or origin, these reference are not relevant anymore. And merging is the goal of the procurement group.

> commit 2: Well, it is just to give a direct link to the sale order which is missing (the origin is nice, but honestly a click is fast and easy if there is only one relation. So this commit is giving a solution...matter of taste, but I dislike the origin as only source of relation and the chatter is packed with not useless, but flooded messages about reserve, unreserve, reserve, unreserve, etc. of stock move states.

We'll try to think of something for master (keeping all the references would be nice). We won't change that in v11 (the origin + the link in the chatter are enough).

> commit 3: This one is simple...you will have a picking in the todo list although there is not a single reservation...this is why I had to fix this for my people in the warehouse...useless information in the ToDo list.

Maybe there's a bug, but how can i reproduce it?

> commit 4: not convinced? hmm, you should come to the two warehouses having a lot of information and you will be immediately convinced that this is a life saver ;-)

We noted the feedback, let's wait for an answer about this topic.

**Comment by wtaferner** [comment]
> I prefer to not make a patch for a potential corner case. And, as i explained, as soon as a different delivery is merged with a different partner or origin, these reference are not relevant anymore. And merging is the goal of the procurement group.

Hmm, a pity as I think I was not doing it without a reason...sorry that I can't nail it by an argument.

> We'll try to think of something for master (keeping all the references would be nice). We won't change that in v11 (the origin + the link in the chatter are enough).

Well, the link in the chatter is lost in the mass of messages, so unusable in this case, but ok, not a big deal IMHO as not relevant for stable (or at least nothing to argue about)

> Maybe there's a bug, but how can i reproduce it?

It is a bug and this is the fix for it ;-) Reproduction...reserve and unreserve...then you will have it in status ready although nothing is reserved. Not sure if I tried on runbot, but I guess it should be enough for the status waiting that no ```stock.move.line``` is there, right?

> We noted the feedback, let's wait for an answer about this topic.

Ok, I will wait patiently for the answer and your decision...so basically commit 1 + 3 are things which should be merged in stable the rest is nice to have IMHO although most of them should not be too late for the fresh 11.0.

**Comment by sle-odoo** [comment]
> Hmm, a pity as I think I was not doing it without a reason...sorry that I can't nail it by an argument.

As soon as you remember the bug and i can reproduce it, i'll consider the patch :)

>[...]reserve and unreserve...then you will have it in status ready although nothing is reserved.[...]

Nope, at least no so simple, after an unreserve the picking goes back to waiting. I played a little with put in pack and move lines but a qty done set but everything seems to work as expected.

**Comment by sle-odoo** [comment]
@wtaferner After discussing it with @lap-odoo, we decided we won't change the origin field to use a m2o as you suggested. We plan to improve the "merge" of picking, eg support batch picking as well as group picking, and according to the user's choice the origin information should be kept or not (to all the references, so i guess we would have to use a proper o2m this time).

As we talked about all the points of this PR i'm closing it, but sure if you remember how to reproduce your bug please report it somewhere and we'll fix it.

Thanks,
Simon

**Comment by wtaferner** [comment]
@sle-odoo 
Hmm, closing the whole without taking action is a bit difficult...I think you missed the important points...the Boolean will destroy clean relations for the partner and the not reproducible (for you) wrong state seems to be an issue of Odoo core...please reopen and give me at least the time to give you a clearer reproduction...reading the code I have changed is obviously not enough to show the issue ;-)

**Pull request closed without merging**